crates/durability/src/lib.rs

```diff
@@ -0,0 +1,101 @@
+use std::sync::Arc;
+
+pub use spacetimedb_commitlog::{error, payload::Txdata, Decoder, Transaction};
+
+mod imp;
+pub use imp::{local, Local};
+
+/// Transaction offset.
+///
+/// The transaction offset is essentially a monotonic counter of all
+/// transactions submitted to the durability layer, starting from zero.
+///
+/// While the implementation may not guarantee that the sequence contains no
+/// gaps, it must guarantee that a higher transaction offset implies durability
+/// of all offsets smaller than it.
+pub type TxOffset = u64;
+
+/// The durability API.
+///
+/// NOTE: This is a preliminary definition, still under consideration.
+///
+/// A durability implementation accepts a payload representing a single database
+/// transaction via [`Durability::append_tx`] in a non-blocking fashion. The
+/// payload _should_ become durable eventually. [`TxOffset`]s reported by
+/// [`Durability::durable_tx_offset`] shall be considered durable to the
+/// extent the implementation can guarantee.
+pub trait Durability: Send + Sync {
+    /// The payload representing a single transaction.
+    type TxData;
+
+    /// Submit the transaction payload to be made durable.
+    ///
+    /// This method must never block, and accept new transactions even if they
+    /// cannot be made durable immediately.
+    ///
+    /// A permanent failure of the durable storage may be signalled by panicking.
+    fn append_tx(&self, tx: Self::TxData);
+
+    /// The [`TxOffset`] considered durable.
+    ///
+    /// A `None` return value indicates that the durable offset is not known,
+    /// either because nothing has been persisted yet, or because the status
+    /// cannot be retrieved.
+    fn durable_tx_offset(&self) -> Option<TxOffset>;
```

Comment:
We could consider adding an observer pattern here if we feel we need to be notified immediately whenever the `durable_tx_offset` changes. Depends on the implementation. I pulled it out of Harald's thing because it was overly complicated and OmniPaxos works by polling anyway.